fastcrypto-zkp/src/bn254/zk_login_api.rs

```diff
@@ -0,0 +1,164 @@
+                .to_string(),
+            "4082367875863433681332203403145435568316851327593401208105741076214120093531"
+                .to_string(),
+        ],
+        vec!["1".to_string(), "0".to_string()],
+    ]);
+
+    // Create a vector of G1Affine elements from the IC
+    let mut vk_gamma_abc_g1 = Vec::new();
+    for e in vec![
+        vec![
+            "7601221783497382045435100727010102844416767995017297605284115099608422303035"
+                .to_string(),
+            "8749785198598536603958085261928419291825402152367782685067088145065090991309"
+                .to_string(),
+            "1".to_string(),
+        ],
+        vec![
+            "2844107402968053321142842260538249836495213364133637503989930436252095154777"
+                .to_string(),
+            "6671443994502368977962577284247390754840595243304510253358092664535353826787"
+                .to_string(),
+            "1".to_string(),
+        ],
+    ] {
+        let g1 = g1_affine_from_str_projective(e);
+        vk_gamma_abc_g1.push(g1);
+    }
+
+    let vk = VerifyingKey {
+        alpha_g1: vk_alpha_1,
+        beta_g2: vk_beta_2,
+        gamma_g2: vk_gamma_2,
+        delta_g2: vk_delta_2,
+        gamma_abc_g1: vk_gamma_abc_g1,
+    };
+
+    // Conver thte verifying key into the prepared form.
+    process_vk_special(&Bn254VerifyingKey(vk)).as_arkworks_pvk()
+}
+
+/// Entry point for the ZkLogin API.
+pub fn verify_zk_login(
+    input: &ZkLoginInputs,
+    max_epoch: u64,
+    eph_pubkey_bytes: &[u8],
+    all_jwk: &HashMap<(String, String), JWK>,
+    usage: Environment,
+) -> Result<(), FastCryptoError> {
+    // Load the expected JWK based on (kid, iss).
+    let jwk = all_jwk
+        .get(&(input.get_kid().to_string(), input.get_iss().to_string()))
+        .ok_or_else(|| FastCryptoError::GeneralError("JWK not found".to_string()))?;
+
+    // Decode modulus to bytes.
+    let modulus = Base64UrlUnpadded::decode_vec(&jwk.n).map_err(|_| {
+        FastCryptoError::GeneralError("Invalid Base64 encoded jwk modulus".to_string())
+    })?;
+
+    // Calculat all inputs hash and passed to the verification function.
+    match verify_zk_login_proof_with_fixed_vk(
+        usage,
+        input.get_proof().as_arkworks(),
+        &input.calculate_all_inputs_hash(eph_pubkey_bytes, &modulus, max_epoch)?,
+    ) {
+        Ok(true) => Ok(()),
+        Ok(false) | Err(_) => Err(FastCryptoError::GeneralError(
+            "Groth16 proof verify failed".to_string(),
+        )),
+    }
+}
+
+/// Verify a proof against its public inputs using the fixed verifying key.
+fn verify_zk_login_proof_with_fixed_vk(
+    usage: Environment,
+    proof: Proof<Bn254>,
+    public_inputs: &[Bn254Fr],
+) -> Result<bool, FastCryptoError> {
+    let pvk = match usage {
+        Environment::Production => &GLOBAL_VERIFYING_KEY,
+        Environment::Test => &INSECURE_GLOBAL_VERIFYING_KEY,
+    };
+    Groth16::<Bn254>::verify_with_processed_vk(pvk, public_inputs, &proof)
```

Comment:
You can avoid the map_err because the verify function also returns a FastCryptoError.

Reply:
this returns a arkworks error enum so needs to be mapped